File: advent9/advent9.dart
import 'dart:io';
import 'dart:core';
bool touching(List<int> head_position,List<int> tail_position) {
	if ( head_position[0] - tail_position[0] <=1 
			&& head_position[0]-tail_position[0]>=-1){
		if ( head_position[1] - tail_position[1] <=1 
				&& head_position[1]-tail_position[1]>=-1){
			return true;

		}
	}
	return false;
}

List<int> step(String s, List<int> position) {
	if (s=='U') {
		position[0]+=1;
	}
	if (s=='D') {
		position[0]-=1;
	}
	if (s=='R') {
		position[1]+=1;
	}
	if (s=='L') {
		position[1]-=1;
	}
	return position;
}

List<int> step_tail( List<int> head_position , List<int> tail_position ) {
	int head_row = head_position[0];
	int  head_column = head_position[1];
	int tail_row = tail_position[0];
	int tail_column = tail_position[1];
	bool same_row = (head_row==tail_row);
	bool same_column = (head_column==tail_column);

	if (same_row && ! same_column){
		if (head_column>tail_column){
			tail_position[1]++;
		}
		else{
			tail_position[1]--;
		}
	}
	else if (same_column && ! same_row) {
		if (head_row > tail_row){
			tail_position[0]++;
		}
		else{
			tail_position[0]--;
		}
	}
	else if ( ! same_column && ! same_row ) {
		if (head_column>tail_column && head_row>tail_row){
			tail_position[0]++;
			tail_position[1]++;
		}
		else if (head_column>tail_column && head_row<tail_row){
			tail_position[0]--;
			tail_position[1]++;
		}
		else if (head_column<tail_column && head_row>tail_row){
			tail_position[0]++;
			tail_position[1]--;
		}
		else if (head_column<tail_column && head_row<tail_row){
			tail_position[0]--;
			tail_position[1]--;
		}

	}
	return tail_position;
}


void partone(){
	List<String> places = ['0 0'];
	List<int> head_position=[0,0];
	List<int> tail_position=[0,0];

	for ( var line = stdin.readLineSync();line!=null;line = stdin.readLineSync()){
		int distance = int.parse(line.substring(line.indexOf(' ')+1));
		String direction = line.substring(0,1);

		if ( direction == 'D' ) {
			for ( int i = 0 ; i < distance ; i++ ){
				head_position = step('D',head_position);

				if (! touching ( head_position , tail_position ) ) {
					tail_position = step_tail(head_position,tail_position);
					String place = tail_position[0].toString() + ' ' + tail_position[1].toString();
					var placecheck = places.where((e)=>e==place);
					if (placecheck.length == 0) {
						places.add(place);
					}
				}
			}
		}
		if ( direction == 'U' ) {
			for ( int i = 0 ; i < distance ; i++ ){
				head_position = step('U',head_position);

				if (! touching ( head_position , tail_position ) ) {
					tail_position = step_tail(head_position,tail_position);
					String place = tail_position[0].toString() + ' ' + tail_position[1].toString();
					var placecheck = places.where((e)=>e==place);
					if (placecheck.length == 0) {
						places.add(place);
					}
				}
			}
		}
		if ( direction == 'R' ) {
			for ( int i = 0 ; i < distance ; i++ ){
				head_position = step('R',head_position);

				if (! touching ( head_position , tail_position ) ) {
					tail_position = step_tail(head_position,tail_position);
					String place = tail_position[0].toString() + ' ' + tail_position[1].toString();
					var placecheck = places.where((e)=>e==place);
					if (placecheck.length == 0) {
						places.add(place);
					}
				}
			}
		}
		if ( direction == 'L' ) {
			for ( int i = 0 ; i < distance ; i++ ){
				head_position = step('L',head_position);

				if (! touching ( head_position , tail_position ) ) {
					tail_position = step_tail(head_position,tail_position);
					String place = tail_position[0].toString() + ' ' + tail_position[1].toString();
					var placecheck = places.where((e)=>e==place);
					if (placecheck.length == 0) {
						places.add(place);
					}
				}
			}
		}
	}
	print(places.length);
}

void parttwo(){
	List<int> head_position = [0,0];
		List<int> two = [0,0];
		List<int> three = [0,0];
		List<int> four = [0,0];
		List<int> five = [0,0];
		List<int> six = [0,0];
		List<int> seven = [0,0];
		List<int> eight = [0,0];
		List<int> nine = [0,0];
		List<int> tail_position = [0,0];

		List<List<int>> knots = [head_position,two,three,four,five,six,seven,eight,nine,tail_position];

		List<String> places2 = [];


		for(var line = stdin.readLineSync() ; line != null ; line = stdin.readLineSync()){
			String direction = line.substring(0,line.indexOf(' '));
			int distance = int.parse(line.substring(line.indexOf(' ')+1));
			if ( direction == 'U' ){
				for (int i = 0 ; i < distance ; i++ ) {
				head_position = step('U',head_position);
				for (int j = 0 ; j < 10 ; j++ ) {
					if ( ! touching(knots[i],knots[i+1]) ){
						knots[i+1] = step_tail(knots[i],knots[i+1]);
					}

				}

				
					var check = places2.where((e)=>e==tail_position[0].toString()+' '+tail_position[1].toString());
					if (check.length == 0) {
						places2.add(tail_position[0].toString()+' '+tail_position[1].toString());
				}	}
			}
			if ( direction == 'D' ){
				for (int i = 0 ; i < distance ; i++ ) {
				head_position = step('D',head_position);
				for (int j = 0 ; j < 10 ; j++ ) {
					if ( ! touching(knots[i],knots[i+1]) ){
						knots[i+1] = step_tail(knots[i],knots[i+1]);
					}

				}

				
					var check = places2.where((e)=>e==tail_position[0].toString()+' '+tail_position[1].toString());
					if (check.length == 0) {
						places2.add(tail_position[0].toString()+' '+tail_position[1].toString());
				}	}
			}
			if ( direction == 'R' ){
				for (int i = 0 ; i < distance ; i++ ) {
				head_position = step('R',head_position);
				for (int j = 0 ; j < 10 ; j++ ) {
					if ( ! touching(knots[i],knots[i+1]) ){
						knots[i+1] = step_tail(knots[i],knots[i+1]);
					}

				}

				
					var check = places2.where((e)=>e==tail_position[0].toString()+' '+tail_position[1].toString());
					if (check.length == 0) {
						places2.add(tail_position[0].toString()+' '+tail_position[1].toString());
				}	}
			}
			if ( direction == 'L' ){
				for (int i = 0 ; i < distance ; i++ ) {
				head_position = step('L',head_position);
				for (int j = 0 ; j < 10 ; j++ ) {
					if ( ! touching(knots[i],knots[i+1]) ){
						knots[i+1] = step_tail(knots[i],knots[i+1]);
					}

				}

				
					var check = places2.where((e)=>e==tail_position[0].toString()+' '+tail_position[1].toString());
					if (check.length == 0) {
						places2.add(tail_position[0].toString()+' '+tail_position[1].toString());
				}	}
			}
		}

		print(places2.length);
}

main(){
	partone();
	parttwo();
}

File: advent9/advent9.py
# positions = [] 

# row = [] 
# for j in range(100):
#     for i in range(100) : 
#         row.append('o')
#     positions.append(row)


def touching(head_position,tail_position) -> bool :
    if head_position[0] - tail_position[0] <= 1 and head_position[0] - tail_position[0] >= -1 :
        if head_position[1] - tail_position[1] <= 1 and head_position[1] - tail_position[1] >= -1 :
            return True
    return False


def step(s:str,pos):
    if s == 'U' :
        pos[0] += 1
    if s == 'D' :
        pos[0] -= 1
    if s == 'R' :
        pos[1] += 1
    if s == 'L' :
        pos[1] -= 1
    return pos

#   12345
# 1 nnnnn
# 2 naaan
# 3 natan
# 4 naaan
# 5 nnnnn
def step_tail(head_position,tail_position):
    head_row = head_position[0]
    head_column = head_position[1]
    tail_row = tail_position[0]
    tail_column = tail_position[1]
    same_row = (head_row==tail_row)
    same_column = (head_column==tail_column)
    if same_row and not same_column :
        if head_column > tail_column :
            tail_position[1]+=1
        else :
            tail_position[1]-=1
    elif same_column and not same_row :
        if head_row > tail_row :
            tail_position[0]+=1
        else :
            tail_position[0]-=1
    elif not same_column and not same_row :

        if head_column > tail_column and head_row > tail_row :
            tail_position[0]+=1
            tail_position[1]+=1

        elif head_column > tail_column and head_row < tail_row:
            tail_position[0]-=1
            tail_position[1]+=1


        elif head_column < tail_column and head_row > tail_row:
            tail_position[0]+=1
            tail_position[1]-=1

        elif head_column < tail_column and head_row < tail_row:
            tail_position[0]-=1
            tail_position[1]-=1
    return tail_position



def parttwo():


    head_position = [0,0]
    two = [0,0]
    three = [0,0]
    four = [0,0]
    five = [0,0]
    six = [0,0]
    seven = [0,0]
    eight = [0,0]
    nine = [0,0]
    tail_position = [0,0]

    knots = [head_position,two,three,four,five,six,seven,eight,nine,tail_position]

    places2 = []
    with open('input.txt','r',encoding='utf-8') as r :
        for i in r:
            distance: int = int(i[i.index(' ')+1:])
            # print(distance)
            if 'R' in str(i):
                for k in range(distance):
                    head_position = step('R',head_position)
                    for j in range(0,9):
                        if not touching(knots[j],knots[j+1]):
                            knots[j+1] = step_tail(knots[j],knots[j+1])
                    if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places2:
                        places2.append(str(tail_position[0]) + ' ' + str(tail_position[1]))
            if 'L' in str(i):
                for k in range(distance):
                    head_position = step('L',head_position)
                    for j in range(0,9):
                        if not touching(knots[j],knots[j+1]):
                            knots[j+1] = step_tail(knots[j],knots[j+1])
                    if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places2:
                        places2.append(str(tail_position[0]) + ' ' + str(tail_position[1]))
            if 'U' in str(i):
                for k in range(distance):
                    head_position = step('U',head_position)
                    for j in range(0,9):
                        # print(j)
                        if not touching(knots[j],knots[j+1]):
                            knots[j+1] = step_tail(knots[j],knots[j+1])
                    if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places2:
                        places2.append(str(tail_position[0]) + ' ' + str(tail_position[1]))
            if 'D' in str(i):
                for k in range(distance):
                    head_position = step('D',head_position)
                    for j in range(0,9):
                        if not touching(knots[j],knots[j+1]):
                            knots[j+1] = step_tail(knots[j],knots[j+1])
                    if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places2:
                        places2.append(str(tail_position[0]) + ' ' + str(tail_position[1]))

    print(len(places2))


# R = [0,0]->[0,1] == position[1]++
# L = [0,0]->[0,-1 == position[1]--
# U = [0,0]->[1,0] == position[0]++
# D = [0,0]->[-1,0]== position[0]--
def partone():
    places = ['0 0']
    head_position=[0,0]
    tail_position=[0,0]
    with open('input.txt','r',encoding='utf-8') as r :
        for i in r:
            # print(head_position)
            # print(tail_position)
            # print (i)
            distance: int = int(i[i.index(' ')+1:])
            # print(distance)
            if 'R' in str(i):
                for i in range(distance):
                    head_position = step('R',head_position)
                    if not touching(head_position,tail_position):
                        # print('fffffffff')
                        tail_position = step_tail(head_position,tail_position)
                        # print(tail_position)
                        if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places:
                            places.append(str(tail_position[0]) + ' ' + str(tail_position[1]))

            if 'L' in str(i):
                for i in range(distance):
                    head_position = step('L',head_position)
                    if not touching(head_position,tail_position):
                        tail_position = step_tail(head_position,tail_position)
                        if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places:
                            places.append(str(tail_position[0]) + ' ' + str(tail_position[1]))

            if 'U' in str(i):
                for i in range(distance):
                    head_position = step('U',head_position)
                    if not touching(head_position,tail_position):
                        tail_position = step_tail(head_position,tail_position)
                        if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places:
                            places.append(str(tail_position[0]) + ' ' + str(tail_position[1]))

            if 'D' in str(i):
                for i in range(distance):
                    head_position = step('D',head_position)
                    if not touching(head_position,tail_position):
                        tail_position = step_tail(head_position,tail_position)
                        if str(tail_position[0]) + ' ' + str(tail_position[1]) not in places:
                            places.append(str(tail_position[0]) + ' ' + str(tail_position[1]))

                 
        print(len(places))

def main():
    partone()
    parttwo()

if __name__ == '__main__':
    main()


